### DISCUSSION TITLE
OpenSatKit Docker Container

### DISCUSSION THREAD

**Discussion by kestr31** [OpenSatKit/OpenSatKit#103] [General]
Hi, many people are experiencing difficulties installing OpenSatKit with COSMOS due to outdated dependencies. Demo version of COSMOS is based on QT4 which isn't supported on Ubuntu distributions released later on.

Although Ubuntu 20.04 has a private repository which provides QT4, this issue can go worse as time passes by. Soi, I thought that using a container might prevent this issue since container can easily run a linux distro different from the host.

I made a small project named [Docker-OpenSatKit](https://github.com/kestr31/Docker-OpenSatKit) and this might help people avoiding deprecated dependency issues. I hope this can help the others trying to use OpenSatKit for their cubesat development.

Again, thank you for the great work of OpenSatKit. This really helps developing and simulating a cubesat system.

**Comment by dmccomas** [comment]
I greatly appreciate you doing this work and sharing it. Using
containers has been on my mind as a solution but I'm a container novice and
didn't know how to do the user interface. I'm looking forward to trying it
out.

**Comment by JiveyGuy** [comment]
This is awesome, I have a few questions:

- Can this use external repository references so we get the most up-to-date OpenSatKit even years down the road?
- Are there plans to make platform-independent install steps? 
    -  Needing a Windows host with WSL2 makes this hard for teams that already have a bunch of work tailor-made for Linux.

**Reply by kestr31** [reply]
Thank you. About the questions, my answers are as following:


> Can this use external repository references so we get the most up-to-date OpenSatKit even years down the road?

- Sure, you can check the line 100 of the `Dockerfile`.

> Are there plans to make platform-independent install steps?
> Needing a Windows host with WSL2 makes this hard for teams that already have a bunch of work tailor-made for Linux.

- Well actually, it is already done. WSL2 is basically just a linux distro that runs on Hyper-V(Although it is slightly different on the lower level, still, we can regard it just as other linux environments on most use cases).
- That is, **you can run the container anywhere if that system has a recent version of AMD64 Linux Kernel**.
- Only difference that occurs on WSL2 is additional environment variable / volume / device settings for the [WSLg](https://github.com/microsoft/wslg).
  - Of course, in order to use WSLg inside a container, [the recent version  of mesa](https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa) is required. It is included in the container.
  - Without installing it, frame will significantly drop on the WSL environment since the old mesa does not support WSLg.


I hope this helped.

**Reply by JiveyGuy** [reply]
Thank you for the reply!
